src/main/java/io/cryostat/net/TargetConnectionManager.java

```diff
@@ -164,16 +176,80 @@ private JFRConnection attemptConnectAsHostPortPair(ConnectionDescriptor connecti
     private JFRConnection connect(
             ConnectionDescriptor cacheKey, JMXServiceURL url, Optional<Credentials> credentials)
             throws Exception {
+        JMXConnectionOpened evtOpened = new JMXConnectionOpened(url.toString());
+        JMXConnectionFailed evtFailed = new JMXConnectionFailed(url.toString());
+
         logger.info("Creating connection for {}", url.toString());
-        return jfrConnectionToolkit
-                .get()
-                .connect(
-                        url,
-                        credentials.orElse(null),
-                        Collections.singletonList(() -> this.connections.invalidate(cacheKey)));
+        evtOpened.begin();
+        evtFailed.begin();
+        try {
+            JFRConnection connection =
+                    jfrConnectionToolkit
+                            .get()
+                            .connect(
+                                    url,
+                                    credentials.orElse(null),
+                                    Collections.singletonList(
+                                            () -> this.connections.invalidate(cacheKey)));
+            evtOpened.end();
+            evtFailed.end();
+            if (evtOpened.shouldCommit()) {
```

Comment:
I would move this into the `finally` block, so that the Opened event is always recorded even if the connection open actually fails. Then, in the exception handler block, either information about the exception can be optionally included in the Opened event, or the Failed event you've created can be used instead. I think in that case the Failed event doesn't need to be begin()/end()ed - it can just be an instantaneous event that occurs at the end of the timespan of an Opened event.

However, to be more consistent, I think including information about the exception in the Opened event might be the better way to go. This would better match @jan-law 's recent event addition(s).

Reply:
I followed what @jan-law did and added an `exceptionThrown` boolean field to my `JMXConnectionOpened` event but on second thought, would a boolean named `connectionFailed` be more appropriate?

Reply:
I don't think it matters too much - if the connection failed it's because an exception was thrown, so they seem pretty equally valid to me. Just for consistency's sake again I would think leaving it as `exceptionThrown` is perhaps better.